File: dbll.c
// Singly Circular Linkedlist

# include<stdio.h>
# include<stdlib.h>
# pragma pack(1)

struct node 
{
    int data;
    struct node* next;
    struct node* prev;
};

typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

void InsertFirst(PPNODE First,PPNODE Last,int no)
{
  PNODE newn = (PNODE)malloc(sizeof(NODE));
  newn ->data = no;
  newn ->next = NULL;
  newn -> prev = NULL;

  if((*First) == NULL && (*Last) == NULL)
  {
    (*First) = (*Last) = newn;
    (*Last) ->next = *First;
  }  
  else
  {
    newn ->next = *First;
    (*First) ->prev = newn;
    *First = newn;
     (*Last) ->next = *First;
  }
   
}

void InsertLast(PPNODE First,PPNODE Last,int no)
{
    PNODE newn = (PNODE)malloc(sizeof(NODE));
  newn ->data = no;
  newn ->next = NULL;
  newn -> prev = NULL;

  if((*First) == NULL && (*Last) == NULL)
  {
    (*First) = (*Last) = newn;
    (*Last) ->next = *First;
  }  

  else
  {
    PNODE Temp = *First;
    PNODE temp1 = NULL;
    do
    {
      Temp = Temp ->next;
    }while(Temp != (*Last)->next);
    (*Last) ->next = newn;
    newn ->prev = *Last;
    *Last = (*Last)->next ;
    (*Last)->next = *First;
  }
}
void Display(PNODE First,PNODE Last)
{
    PNODE Temp = First;
    do
    {
        printf("|%d|<=>",Temp->data);
        Temp = Temp ->next;
    }while(Temp != Last->next);
    printf("    \n");
}

void DeleteFirst(PPNODE First,PPNODE Last)
{
    if(*First == NULL || *Last == NULL)
    {
        printf("Empty LinkedList \n");
        return;
    }
    else if((*First) ->next == NULL)
    {
        free(*First);
        *First = NULL;
        *Last = NULL;
    }
    else
    {
        PNODE Temp = *First;
        *First = (*First)->next;
        (*First) ->prev = NULL;
        free(Temp);
        (*Last)->next = *First;
    }
}
void DeleteLast(PPNODE First,PPNODE Last)
{
    if(*First == NULL || *Last == NULL)
    {
        printf("Empty LinkedList \n");
        return;
    }
    else if((*First) ->next == NULL)
    {
        free(*First);
        *First = NULL;
        *Last = NULL;
    }
    else
    {
        PNODE temp = *Last;
        *Last = (*Last) ->prev;
        free(temp);
        (*Last) ->next =*First;
    }
}

int Count (PNODE First,PNODE Last)
{
    int iCnt = 0;
    PNODE temp = First;
    do
    {
    
        iCnt++;
        temp = temp->next;
    }while(temp != Last->next);
    return iCnt;
}
void InsertAtPos(PPNODE First,PPNODE Last,int no,int iPos)
{
    int iCount = Count(*First,*Last);

    PNODE newn = (PNODE)malloc(sizeof(NODE));
    newn ->data = no;
    newn -> next = NULL;
    newn -> prev = NULL;

    if(iPos < 1 || iPos > iCount+1)
    {   
        printf("intvalid Postion");
        return;
    } 
    else if(iPos == 1)
    {
        InsertFirst(First,Last,no);
    }
    else if(iPos == iCount+1)
    {
        InsertLast(First,Last,no);
    }
    else
    {
        PNODE temp = *First;
        int no = 0;
        do
        {
            no++;
            temp = temp->next;
        }while(no != iPos-1);
    PNODE Temp1 = temp ->next;
    newn ->next = Temp1;
    newn ->prev = temp;
    (*Last)->next = *First;
    }
}

void DeleteAtPose(PPNODE First,PPNODE Last,int iPos)
{
    int iCount = Count(*First,*Last);
    if(iPos < 1 || iPos > iCount+1)
    {   
        printf("intvalid Postion");
        return;
    } 
    else if(iPos == 1)
    {
        DeleteFirst(First,Last);
    }
    else if(iPos == iCount+1)
    {
        DeleteLast(First,Last);
    }
    else
    {
        PNODE temp = *First;
        for(int i = 0;i<iPos-2;i++)
        {
            temp = temp -> next;
        }
        PNODE Temp1 = temp->next;
        temp ->next = Temp1->next;
        Temp1 ->next ->prev = temp;
        free(Temp1);
        (*Last)->next = *First;      
    }
}
int main()
{
    PNODE Head = NULL;
    PNODE Tail = NULL;

    InsertFirst(&Head,&Tail,51);
    InsertFirst(&Head,&Tail,21);
    InsertFirst(&Head,&Tail,11);
    Display(Head,Tail);

    InsertLast(&Head,&Tail,101);
    InsertLast(&Head,&Tail,111);
    InsertLast(&Head,&Tail,121);
    Display(Head,Tail);

    DeleteFirst(&Head,&Tail);
    Display(Head,Tail);

    DeleteLast(&Head,&Tail);
    Display(Head,Tail);

    InsertAtPos(&Head,&Tail,105,5);
    Display(Head,Tail);
    DeleteAtPose(&Head,&Tail,2);
    Display(Head,Tail);

    return 0;
}

File: dbll.java
// doubly Linear Linked list in java

import java.util.*;

class dbll 
{
    public static void main(String arg[])
    {
        doublyLL dobj = new doublyLL();

        dobj.InsertFirst(51);
        dobj.InsertFirst(21);
        dobj.InsertFirst(11);
        dobj.Display();

        dobj.InsertLast(101);
        dobj.InsertLast(111);
        dobj.InsertLast(121);
        dobj.Display();

        dobj.DeleteFirst();
        dobj.Display();

        dobj.DeleteLast();
        dobj.Display();

    }
}
class Node
{
    public int data;
    public Node next;
    public Node prev;

    public Node(int no)
    {
        this.data = no;
        this.next = null;
        this.prev = null;
    }
}
class doublyLL
{
    private int Count;
    private Node Head;
    private Node Tail;

    public doublyLL()
    {
        this.Count = 0;
        this.Head = null;
        this.Tail = null; 
    }

    public void InsertFirst(int no)
    {
        Node newn = new Node(no);

        if(Head == null && Tail == null)
        {
            Head = newn;
            Tail = newn;
        }
        else
        {
            newn.next = Head;
            Head.prev = newn;
            Head = newn;
        }
        this.Count++;
    }
    public void InsertLast(int no)
    {
        Node newn = new Node(no);

        if(Head == null && Tail == null)
        {
            Head = newn;
            Tail = newn;
        }
        else
        {
            Tail.next = newn;
            newn.prev = Tail;
            Tail= Tail.next;
        }
        this.Count++;
    }
    public void InsertAtPos(int no, int iPos)
    {
        Node newn = new Node(no);
        if(iPos <1 || iPos > Count+1)
        {
            return;
        }
        else if(iPos == 1)
        {
            InsertFirst(no);
        }
        else if(iPos>Count)
        {
            InsertLast(no);
        }
        else
        {
            Node temp = Head;
            for(int i =1;i<iPos-1;i++)
            {
                temp = temp.next;
            }
            Node temp1 = temp.next;
            newn.next = temp1;
            temp1.prev = newn;
            temp.next = newn;
            newn.prev = temp;
        }
        this.Count++;
    }
    public void DeleteFirst()
    {
        if(Count ==0)
        {
            return;
        }
        else if(Count == 1)
        {
            Head = null;
            Tail = null;
        }
        else
        {
            Head = Head.next;
            Head.prev = null;
        }
        this.Count--;
    }
    public void DeleteLast()
    {
        if(Count ==0)
        {
            return;
        }
        else if(Count == 1)
        {
            Head = null;
            Tail = null;
        }
        else
        {
            Node temp = Head;
        Tail=Tail.prev;
        Tail.next= null;
        this.Count--;
        }
        
    }
    public void DeleteAtPos(int iPos)
    {
        if(iPos <0 || iPos>Count)
        {
            return;
        }
        else if(iPos == 1)
        {
            DeleteFirst();
        }
        else if (iPos == Count)
        {
            DeleteLast();
        }
        else
        {
            Node temp = Head;
            for(int i=1;i<iPos-1;iPos++)
            {
                temp = temp.next;
            }
            Node temp1 = temp.next;
            temp.next=temp1.next;
            temp1.next.prev = temp;
            this.Count--;
        }
    }
    public void Display()
    {
        Node temp =Head;
        while(temp != null)
        {
            System.out.print("|"+temp.data+"|->");
            temp=temp.next;
        }
        System.out.println("null");
    }
    public int count()
    {
        return this.Count;
    }
}
